tests/apis/coinbase/websocket.rs

```diff
@@ -1,28 +1,59 @@
-use futures::StreamExt;
 use openlimits::coinbase::{
-    client::websocket::CoinbaseWebsocket,
-    model::websocket::{ChannelType, Subscription},
+    client::websocket::CoinbaseWebsocket, model::websocket::CoinbaseSubscription,
+    CoinbaseParameters,
 };
+use openlimits::exchange_ws::ExchangeWs;
+use openlimits::model::websocket::{OpenLimitsWebSocketMessage, WebSocketResponse};
+use std::sync::mpsc::sync_channel;
 
-#[tokio::test]
-async fn aggregate_trade() {
-    let mut websocket = CoinbaseWebsocket::new("wss://ws-feed.pro.coinbase.com");
+async fn test_subscription_callback(
+    websocket: CoinbaseWebsocket,
+    sub: CoinbaseSubscription,
+    expected_messages: Vec<OpenLimitsWebSocketMessage>,
+) {
+    let (tx, rx) = sync_channel(0);
 
-    let sub = Subscription {
-        channels: vec![ChannelType::Level2],
-        product_ids: vec!["BTC-USD".to_string()],
-    };
+    let mut received_messages: Vec<bool> = expected_messages.iter().map(|_| false).collect();
 
-    websocket.subscribe(sub).await.expect("Couldn't subscribe.");
-
-    websocket
-        .next()
-        .await
-        .expect("Couldn't get next.")
-        .expect("Couldn't get WebSocket message.");
     websocket
-        .next()
+        .subscribe(sub, move |message| {
+            if let Ok(message) = message.as_ref() {
+                if let WebSocketResponse::Generic(message) = message {
+                    let expected_iter = expected_messages.iter().map(|expected| {
+                        std::mem::discriminant(expected) == std::mem::discriminant(&message)
+                    });
+                    for (already_received, currently_received) in
+                        received_messages.iter_mut().zip(expected_iter)
+                    {
+                        if !*already_received {
+                            *already_received = currently_received;
+                        }
+                    }
+                    if received_messages.iter().all(|received| *received) {
+                        tx.send(()).expect("Couldn't send sync message.");
```

Comment:
Any way we can change this to an error somehow?

Reply:
I am introducing a timeout. Because we are waiting to get all the expected messages and the only way this logic fails is if we never get them.